metadata-ingestion-modules/airflow-plugin/src/datahub_airflow_plugin/airflow3/datahub_listener.py

```diff
@@ -0,0 +1,1361 @@
+                )
+
+                # Get token - check airflow.cfg first, then connection password
+                token = conf.get("datahub", "token", fallback=None)
+                if token is None:
+                    token = conn.password
+
+                # Get extra args
+                extra_args = conn.extra_dejson or {}
+
+                logger.debug(
+                    f"Retrieved connection '{conn_id}' from connection API: type={conn_type or 'datahub-rest'}, host={host}, has_token={bool(token)}"
+                )
+
+                return datahub.emitter.rest_emitter.DataHubRestEmitter(
+                    host,
+                    token,
+                    client_mode=ClientMode.INGESTION,
+                    datahub_component="airflow-plugin",
+                    **extra_args,
+                )
+        except Exception as e:
+            logger.debug(
+                f"Failed to create emitter from connection: {e}", exc_info=True
+            )
+            return None
+
+    @property
+    def emitter(self):
+        """Compatibility property that delegates to _get_emitter()."""
+        result = self._get_emitter()
+        if result is None:
+            # Retry emitter creation
+            self._emitter = None  # Reset to force retry
+            return self._get_emitter()
+        return result
+
+    @property
+    def graph(self) -> Optional[DataHubGraph]:
+        if self._graph:
+            return self._graph
+
+        # Use _get_emitter() method to ensure lazy-loading happens first
+        emitter = self._get_emitter()
+        if emitter is not None:
+            import datahub.emitter.rest_emitter
+
+            if isinstance(
+                emitter, datahub.emitter.rest_emitter.DataHubRestEmitter
+            ) and not isinstance(emitter, DataHubGraph):
+                # This is lazy initialized to avoid throwing errors on plugin load.
+                self._graph = emitter.to_graph()
+                self._emitter = self._graph
+
+        return self._graph
+
+    def _set_log_level(self) -> None:
+        """Set the log level for the plugin and its dependencies.
+
+        This may need to be called multiple times, since Airflow sometimes
+        messes with the logging configuration after the plugin is loaded.
+        In particular, the loggers may get changed when the worker starts
+        executing a task.
+        """
+
+        if self.config.log_level:
+            logging.getLogger(__name__.split(".")[0]).setLevel(self.config.log_level)
+        if self.config.debug_emitter:
+            logging.getLogger("datahub.emitter").setLevel(logging.DEBUG)
+
+    def _make_emit_callback(self) -> Callable[[Optional[Exception], str], None]:
+        def emit_callback(err: Optional[Exception], msg: str) -> None:
+            if err:
+                logger.error(f"Error sending metadata to datahub: {msg}", exc_info=err)
+
+        return emit_callback
+
+    def _extract_lineage_from_airflow3(
```

Comment:
I like how precise this function is when it get to logging, but cyclomatic complexity here seems to be a bit too high. Maybe we could introduce helper functions to help reducing that?